File: packages/minting/python/src/lattice_minting/minter.py
# SPDX-License-Identifier: MPL-2.0
"""Minting from a recipe (identity-minting-specification.md §6).

Every result carries a trace of every intermediate value, in the same shape
as the conformance vectors (§10.1), so a mismatch can be located at the
first step that differs."""

from __future__ import annotations

import re
import secrets as _secrets
from collections.abc import Callable, Mapping
from dataclasses import dataclass, field

from . import ucd
from .canonical import encode, hmac_sha256, percent_encode, sha256, tuple_bytes, uuid4_from
from .errors import MintError
from .recipe import Recipe

MAX_POSITION = 9223372036854775807
TOKEN = re.compile(r"[A-Za-z0-9._~-]+")
SLOT = re.compile(r"\{([A-Za-z]+)\}")


def render(template: str, slots: Mapping[str, str]) -> str:
    return SLOT.sub(lambda m: slots[m.group(1)], template)


@dataclass
class Minted:
    iri: str
    claim_iris: list[str] = field(default_factory=list)
    trace: list[dict] = field(default_factory=list)
    canonicalizer: str | None = None


class Minter:
    """Mints IRIs from one verified recipe.

    ``secrets`` maps a claim scheme's key id to the secret's bytes, and is
    needed only for claimed identity. The minter does not retain it beyond
    the calls that use it. ``random_bytes`` returns n cryptographically
    secure random bytes; replace it only in tests.
    """

    def __init__(self, recipe: Recipe, secrets: Mapping[str, bytes] | None = None,
                 random_bytes: Callable[[int], bytes] = _secrets.token_bytes):
        ucd.check_runtime()
        self.recipe = recipe
        self.r = recipe.data
        self._secrets = dict(secrets or {})
        self._random = random_bytes

    # ---- shared steps -------------------------------------------------------------------------------

    def _normalize(self, pipeline: Mapping, raw, index: int, trace: list) -> str:
        if not isinstance(raw, str):
            raise MintError("MissingKeyComponent", f"key component {index} is missing")
        value, after = raw, []
        for step in pipeline["steps"]:
            value = ucd.STEPS[step](value)
            after.append({"step": step, "output": value})
        if value == "":
            raise MintError("EmptyKeyComponent", f"key component {index} is empty after normalization")
        trace.append({"step": "normalize", "component": index, "pipeline": pipeline["id"],
                      "input": raw, "after": after, "output": value})
        return value

    def _key(self, key: Mapping, inputs: Mapping, trace: list) -> tuple[list[str], list[str]]:
        values = list(inputs.get("key") or [])
        if len(values) < len(key["properties"]):
            raise MintError("MissingKeyComponent", f"{len(key['properties'])} key components needed, {len(values)} given")
        scope: list[str] = []
        if key.get("scopeProperty"):
            if not isinstance(inputs.get("scope"), str):
                raise MintError("MissingKeyComponent", "the key's scope value is missing")
            scope = [inputs["scope"]]
        normalized = [self._normalize(key["pipeline"], v, i, trace) for i, v in enumerate(values[: len(key["properties"])])]
        return scope, normalized

    def _digest(self, data: bytes, spec: Mapping, trace: list, mac_key: bytes | None = None,
                key_id: str | None = None, claim: int | None = None) -> str:
        tag = {} if claim is None else {"claim": claim}
        if mac_key is None:
            full = sha256(data)
            trace.append({"step": "digest", **tag, "function": "SHA-256", "hex": full.hex()})
        else:
            full = hmac_sha256(mac_key, data)
            trace.append({"step": "mac", **tag, "function": "HMAC-SHA-256", "keyId": key_id, "hex": full.hex()})
        cut = full[: spec["widthBits"] // 8]
        trace.append({"step": "truncate", **tag, "widthBits": spec["widthBits"], "hex": cut.hex()})
        value = encode(cut, spec["encoding"])
        trace.append({"step": "encode", **tag, "encoding": spec["encoding"], "value": value})
        return value

    def _match(self, pattern: str, value, what: str) -> str:
        if not isinstance(value, str) or re.fullmatch(pattern, value) is None:
            raise MintError("PatternMismatch", f"{what} {value!r} does not match {pattern}")
        return value

    def _uuid(self, inputs: Mapping, trace: list) -> str:
        random_hex = inputs.get("randomHex")
        raw = bytes.fromhex(random_hex) if random_hex is not None else self._random(16)
        value = uuid4_from(raw)
        trace.append({"step": "generate-surrogate", "randomHex": raw.hex(), "value": value})
        return value

    # ---- strategies -----------------------------------------------------------------------------------

    def mint(self, inputs: Mapping) -> Minted:
        """Mint (or, for adopted identity, validate) one identifier.

        ``inputs`` members: ``key`` (list of raw key values), ``scope``,
        ``surrogate`` (caller-supplied surrogates), ``target``, ``epoch``,
        ``seq``, ``namespaceToken`` (position-derived), ``canonicalNQuads``
        and ``canonicalizer`` (content-addressed), ``iri`` (adopted),
        ``randomHex`` (tests only: 16 bytes instead of fresh randomness).
        """
        s = self.r["strategy"]
        trace: list[dict] = []
        if s in ("AdoptedIdentity", "ExternalRegistryIdentity"):
            iri = self._match(self.r["acceptedPattern"], inputs.get("iri"), "identifier")
            trace.append({"step": "validate-iri", "pattern": self.r["acceptedPattern"], "value": iri})
            return Minted(iri, trace=trace)
        if s == "NaturalKeyIdentity":
            scope, normalized = self._key(self.r["key"], inputs, trace)
            parts = [percent_encode(v) for v in scope + normalized]
            trace.append({"step": "percent-encode", "components": parts})
            return self._done(render(self.r["iriTemplate"], {"key": "/".join(parts)}), trace)
        if s == "DerivedHashIdentity":
            scope, normalized = self._key(self.r["key"], inputs, trace)
            components = list(self.r["tuplePrefix"]) + scope + normalized
            data = tuple_bytes(components)
            trace.append({"step": "tuple", "components": components, "hex": data.hex()})
            return self._done(render(self.r["iriTemplate"], {"digest": self._digest(data, self.r["digest"], trace)}), trace)
        if s == "RandomSurrogateIdentity":
            return self._done(render(self.r["iriTemplate"], {"surrogate": self._uuid(inputs, trace)}), trace)
        if s == "SurrogateClaimedIdentity":
            return self._claimed(inputs, trace)
        if s == "PositionDerivedEvent":
            return self._position(inputs, trace)
        if s == "ContentAddressedIdentity":
            return self._content(inputs, trace)
        raise MintError("UnsupportedRecipeFormat", f"strategy {s}")

    def _done(self, iri: str, trace: list, **extra) -> Minted:
        trace.append({"step": "render", "iri": iri})
        return Minted(iri, trace=trace, **extra)

    def _claimed(self, inputs: Mapping, trace: list) -> Minted:
        sur = self.r["surrogate"]
        if sur["kind"] == "CallerSuppliedSurrogate":
            value = self._match(sur["pattern"], inputs.get("surrogate"), "surrogate")
            trace.append({"step": "validate-surrogate", "pattern": sur["pattern"], "value": value})
        else:
            value = self._uuid(inputs, trace)
        iri = render(self.r["iriTemplate"], {"surrogate": value})
        trace.append({"step": "render", "iri": iri})
        claims = self.r["claims"]
        scope, normalized = self._key(claims[0]["key"], inputs, trace)
        claim_iris = []
        for ix, c in enumerate(claims):
            key = self._secrets.get(c["keyId"])
            if not key:
                raise MintError("MissingSecret", f"no secret supplied for key id {c['keyId']!r}")
            components = [c["schemeVersion"], c["constraintId"], scope[0] if scope else "", *normalized]
            data = tuple_bytes(components)
            trace.append({"step": "tuple", "claim": ix, "components": components, "hex": data.hex()})
            mac = self._digest(data, c["mac"], trace, mac_key=key, key_id=c["keyId"], claim=ix)
            claim_iri = render(c["iriTemplate"], {"constraintId": c["constraintId"], "schemeVersion": c["schemeVersion"], "mac": mac})
            trace.append({"step": "render", "claim": ix, "iri": claim_iri})
            claim_iris.append(claim_iri)
        return Minted(iri, claim_iris=claim_iris, trace=trace)

    def _position(self, inputs: Mapping, trace: list) -> Minted:
        ns_spec = self.r["namespace"]
        if ns_spec["derivation"] == "HashedTargetDerivation":
            target = inputs.get("target")
            if not isinstance(target, str) or not target:
                raise MintError("MissingKeyComponent", "the target IRI is missing")
            components = ["occurrence-namespace/1", target]
            data = tuple_bytes(components)
            trace.append({"step": "tuple", "components": components, "hex": data.hex()})
            namespace = self._digest(data, ns_spec["digest"], trace)
        else:
            namespace = inputs.get("namespaceToken")
            if not isinstance(namespace, str) or TOKEN.fullmatch(namespace) is None:
                raise MintError("PatternMismatch", f"namespace token {namespace!r} is not [A-Za-z0-9._~-]+")
            trace.append({"step": "validate-token", "pattern": "^[A-Za-z0-9._~-]+$", "value": namespace})
        padded = []
        for name, width in (("epoch", self.r["epochWidth"]), ("seq", self.r["sequenceWidth"])):
            v = inputs.get(name)
            if isinstance(v, bool) or not isinstance(v, int):
                raise MintError("MissingKeyComponent", f"{name} is missing or not an integer")
            if not 0 <= v <= MAX_POSITION:
                raise MintError("PositionOutOfRange", f"{name} {v} is outside 0..{MAX_POSITION}")
            padded.append(f"{v:0{width}d}")
        trace.append({"step": "pad", "epoch": padded[0], "seq": padded[1]})
        iri = render(self.r["iriTemplate"], {"namespace": namespace, "epoch": padded[0], "seq": padded[1]})
        return self._done(iri, trace)

    def _content(self, inputs: Mapping, trace: list) -> Minted:
        canonicalizer = inputs.get("canonicalizer")
        if not isinstance(canonicalizer, str) or not ucd.trim_white_space(canonicalizer):
            raise MintError("CanonicalizerNotDeclared",
                            "content-addressed minting needs a statement of which RDFC-1.0 implementation produced "
                            "the bytes (identity-minting-specification.md §7, CA-1)")
        nquads = inputs.get("canonicalNQuads")
        if not isinstance(nquads, str):
            raise MintError("MissingKeyComponent", "canonicalNQuads is missing")
        value = self._digest(nquads.encode("utf-8"), self.r["digest"], trace)
        return self._done(render(self.r["iriTemplate"], {"digest": value}), trace, canonicalizer=canonicalizer)

File: packages/minting/java/src/main/java/org/nebularis/lattice/minting/Minter.java
// SPDX-License-Identifier: MPL-2.0
package org.nebularis.lattice.minting;

import static org.nebularis.lattice.minting.Values.entry;
import static org.nebularis.lattice.minting.Values.integer;
import static org.nebularis.lattice.minting.Values.list;
import static org.nebularis.lattice.minting.Values.map;

import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HexFormat;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.IntFunction;
import java.util.function.UnaryOperator;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Mints IRIs from one verified recipe (identity-minting-specification.md §6).
 *
 * <p>{@code secrets} maps a claim scheme's key id to the secret's bytes, and is
 * needed only for claimed identity. The minter copies it at construction and
 * never returns or logs it. {@code randomBytes} returns n cryptographically
 * secure random bytes; replace it only in tests.
 *
 * <p>Request inputs are the members listed in §6: {@code key} (a list of raw
 * key values), {@code scope}, {@code surrogate}, {@code target},
 * {@code namespaceToken}, {@code epoch}, {@code seq}, {@code canonicalNQuads},
 * {@code canonicalizer}, {@code iri}, and {@code randomHex} (vectors only).
 */
public final class Minter {
    public static final long MAX_POSITION = Long.MAX_VALUE;
    private static final String TOKEN_PATTERN = "^[A-Za-z0-9._~-]+$";
    private static final Pattern TOKEN = Pattern.compile(TOKEN_PATTERN);
    private static final Pattern SLOT = Pattern.compile("\\{([A-Za-z]+)\\}");
    private static final SecureRandom RANDOM = new SecureRandom();

    private final Recipe recipe;
    private final Map<String, Object> r;
    private final Map<String, byte[]> secrets;
    private final IntFunction<byte[]> random;

    public Minter(Recipe recipe) {
        this(recipe, Map.of());
    }

    public Minter(Recipe recipe, Map<String, byte[]> secrets) {
        this(recipe, secrets, Minter::secureRandomBytes);
    }

    public Minter(Recipe recipe, Map<String, byte[]> secrets, IntFunction<byte[]> randomBytes) {
        Ucd.checkRuntime();
        this.recipe = Objects.requireNonNull(recipe);
        this.r = recipe.data();
        Map<String, byte[]> copy = new HashMap<>();
        if (secrets != null) {
            secrets.forEach((k, v) -> copy.put(k, v == null ? null : v.clone()));
        }
        this.secrets = copy;
        this.random = Objects.requireNonNull(randomBytes);
    }

    private static byte[] secureRandomBytes(int n) {
        byte[] b = new byte[n];
        RANDOM.nextBytes(b);
        return b;
    }

    public Recipe recipe() {
        return recipe;
    }

    static String render(String template, Map<String, String> slots) {
        return SLOT.matcher(template).replaceAll(m -> {
            String v = slots.get(m.group(1));
            if (v == null) {
                throw new IllegalArgumentException("template slot {" + m.group(1) + "} is not available");
            }
            return Matcher.quoteReplacement(v);
        });
    }

    // ---- shared steps --------------------------------------------------------------------------------

    private String normalize(Map<String, Object> pipeline, Object raw, int index, List<Map<String, Object>> trace) {
        if (!(raw instanceof String input)) {
            throw new MintException(MintError.MissingKeyComponent, "key component " + index + " is missing");
        }
        String value = input;
        List<Object> after = new ArrayList<>();
        for (Object step : list(pipeline.get("steps"))) {
            UnaryOperator<String> fn = Ucd.step((String) step);
            value = fn.apply(value);
            after.add(entry("step", step, "output", value));
        }
        if (value.isEmpty()) {
            throw new MintException(MintError.EmptyKeyComponent, "key component " + index + " is empty after normalization");
        }
        trace.add(entry("step", "normalize", "component", (long) index, "pipeline", pipeline.get("id"),
                "input", input, "after", after, "output", value));
        return value;
    }

    private record KeyValues(List<String> scope, List<String> normalized) {
    }

    private KeyValues key(Map<String, Object> key, Map<String, ?> inputs, List<Map<String, Object>> trace) {
        List<?> values = inputs.get("key") instanceof List<?> l ? l : List.of();
        int n = list(key.get("properties")).size();
        if (values.size() < n) {
            throw new MintException(MintError.MissingKeyComponent, n + " key components needed, " + values.size() + " given");
        }
        List<String> scope = List.of();
        if (key.get("scopeProperty") instanceof String sp && !sp.isEmpty()) {
            if (!(inputs.get("scope") instanceof String s)) {
                throw new MintException(MintError.MissingKeyComponent, "the key's scope value is missing");
            }
            scope = List.of(s);
        }
        List<String> normalized = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            normalized.add(normalize(map(key.get("pipeline")), values.get(i), i, trace));
        }
        return new KeyValues(scope, normalized);
    }

    private String digest(byte[] data, Map<String, Object> spec, List<Map<String, Object>> trace,
                          byte[] macKey, String keyId, Integer claim) {
        byte[] full;
        Map<String, Object> step;
        if (macKey == null) {
            full = Canonical.sha256(data);
            step = entry("step", "digest");
            tag(step, claim);
            step.put("function", "SHA-256");
        } else {
            full = Canonical.hmacSha256(macKey, data);
            step = entry("step", "mac");
            tag(step, claim);
            step.put("function", "HMAC-SHA-256");
            step.put("keyId", keyId);
        }
        step.put("hex", hex(full));
        trace.add(step);
        int width = integer(spec.get("widthBits"));
        byte[] cut = Arrays.copyOf(full, width / 8);
        Map<String, Object> truncate = entry("step", "truncate");
        tag(truncate, claim);
        truncate.put("widthBits", (long) width);
        truncate.put("hex", hex(cut));
        trace.add(truncate);
        String encoding = (String) spec.get("encoding");
        String value = Canonical.encode(cut, encoding);
        Map<String, Object> encode = entry("step", "encode");
        tag(encode, claim);
        encode.put("encoding", encoding);
        encode.put("value", value);
        trace.add(encode);
        return value;
    }

    private static void tag(Map<String, Object> step, Integer claim) {
        if (claim != null) {
            step.put("claim", (long) claim);
        }
    }

    private static String hex(byte[] b) {
        return HexFormat.of().formatHex(b);
    }

    private static String match(String pattern, Object value, String what) {
        if (!(value instanceof String s) || !Pattern.compile(pattern).matcher(s).matches()) {
            throw new MintException(MintError.PatternMismatch, what + " " + value + " does not match " + pattern);
        }
        return s;
    }

    private String uuid(Map<String, ?> inputs, List<Map<String, Object>> trace) {
        byte[] raw = inputs.get("randomHex") instanceof String h ? HexFormat.of().parseHex(h) : random.apply(16);
        String value = Canonical.uuid4From(raw);
        trace.add(entry("step", "generate-surrogate", "randomHex", hex(raw), "value", value));
        return value;
    }

    // ---- strategies ----------------------------------------------------------------------------------

    /** Mint (or, for adopted identity, validate) one identifier. A refusal throws {@link MintException}. */
    public Minted mint(Map<String, ?> inputs) {
        String s = recipe.strategy();
        List<Map<String, Object>> trace = new ArrayList<>();
        return switch (s) {
            case "AdoptedIdentity", "ExternalRegistryIdentity" -> {
                String pattern = (String) r.get("acceptedPattern");
                String iri = match(pattern, inputs.get("iri"), "identifier");
                trace.add(entry("step", "validate-iri", "pattern", pattern, "value", iri));
                yield new Minted(iri, List.of(), trace, null);
            }
            case "NaturalKeyIdentity" -> {
                KeyValues kv = key(map(r.get("key")), inputs, trace);
                List<String> parts = new ArrayList<>();
                for (String v : concat(kv.scope(), kv.normalized())) {
                    parts.add(Canonical.percentEncode(v));
                }
                trace.add(entry("step", "percent-encode", "components", parts));
                yield done(render((String) r.get("iriTemplate"), Map.of("key", String.join("/", parts))), trace, null);
            }
            case "DerivedHashIdentity" -> {
                KeyValues kv = key(map(r.get("key")), inputs, trace);
                List<String> components = new ArrayList<>();
                for (Object p : list(r.get("tuplePrefix"))) {
                    components.add((String) p);
                }
                components.addAll(kv.scope());
                components.addAll(kv.normalized());
                byte[] data = Canonical.tupleBytes(components);
                trace.add(entry("step", "tuple", "components", components, "hex", hex(data)));
                String digest = digest(data, map(r.get("digest")), trace, null, null, null);
                yield done(render((String) r.get("iriTemplate"), Map.of("digest", digest)), trace, null);
            }
            case "RandomSurrogateIdentity" ->
                    done(render((String) r.get("iriTemplate"), Map.of("surrogate", uuid(inputs, trace))), trace, null);
            case "SurrogateClaimedIdentity" -> claimed(inputs, trace);
            case "PositionDerivedEvent" -> position(inputs, trace);
            case "ContentAddressedIdentity" -> content(inputs, trace);
            default -> throw new MintException(MintError.UnsupportedRecipeFormat, "strategy " + s);
        };
    }

    private static List<String> concat(List<String> a, List<String> b) {
        List<String> out = new ArrayList<>(a);
        out.addAll(b);
        return out;
    }

    private static Minted done(String iri, List<Map<String, Object>> trace, String canonicalizer) {
        trace.add(entry("step", "render", "iri", iri));
        return new Minted(iri, List.of(), trace, canonicalizer);
    }

    private Minted claimed(Map<String, ?> inputs, List<Map<String, Object>> trace) {
        Map<String, Object> sur = map(r.get("surrogate"));
        String value;
        if ("CallerSuppliedSurrogate".equals(sur.get("kind"))) {
            String pattern = (String) sur.get("pattern");
            value = match(pattern, inputs.get("surrogate"), "surrogate");
            trace.add(entry("step", "validate-surrogate", "pattern", pattern, "value", value));
        } else {
            value = uuid(inputs, trace);
        }
        String iri = render((String) r.get("iriTemplate"), Map.of("surrogate", value));
        trace.add(entry("step", "render", "iri", iri));
        List<?> claims = list(r.get("claims"));
        KeyValues kv = key(map(map(claims.get(0)).get("key")), inputs, trace);
        List<String> claimIris = new ArrayList<>();
        for (int ix = 0; ix < claims.size(); ix++) {
            Map<String, Object> c = map(claims.get(ix));
            String keyId = (String) c.get("keyId");
            byte[] secret = secrets.get(keyId);
            if (secret == null || secret.length == 0) {
                throw new MintException(MintError.MissingSecret, "no secret supplied for key id " + keyId);
            }
            List<String> components = new ArrayList<>();
            components.add((String) c.get("schemeVersion"));
            components.add((String) c.get("constraintId"));
            components.add(kv.scope().isEmpty() ? "" : kv.scope().get(0));
            components.addAll(kv.normalized());
            byte[] data = Canonical.tupleBytes(components);
            trace.add(entry("step", "tuple", "claim", (long) ix, "components", components, "hex", hex(data)));
            String mac = digest(data, map(c.get("mac")), trace, secret, keyId, ix);
            String claimIri = render((String) c.get("iriTemplate"), Map.of(
                    "constraintId", (String) c.get("constraintId"), "schemeVersion", (String) c.get("schemeVersion"), "mac", mac));
            trace.add(entry("step", "render", "claim", (long) ix, "iri", claimIri));
            claimIris.add(claimIri);
        }
        return new Minted(iri, claimIris, trace, null);
    }

    private Minted position(Map<String, ?> inputs, List<Map<String, Object>> trace) {
        Map<String, Object> ns = map(r.get("namespace"));
        String namespace;
        if ("HashedTargetDerivation".equals(ns.get("derivation"))) {
            if (!(inputs.get("target") instanceof String target) || target.isEmpty()) {
                throw new MintException(MintError.MissingKeyComponent, "the target IRI is missing");
            }
            List<String> components = List.of("occurrence-namespace/1", target);
            byte[] data = Canonical.tupleBytes(components);
            trace.add(entry("step", "tuple", "components", components, "hex", hex(data)));
            namespace = digest(data, map(ns.get("digest")), trace, null, null, null);
        } else {
            if (!(inputs.get("namespaceToken") instanceof String token) || !TOKEN.matcher(token).matches()) {
                throw new MintException(MintError.PatternMismatch,
                        "namespace token " + inputs.get("namespaceToken") + " is not [A-Za-z0-9._~-]+");
            }
            namespace = token;
            trace.add(entry("step", "validate-token", "pattern", TOKEN_PATTERN, "value", namespace));
        }
        String epoch = pad(inputs.get("epoch"), "epoch", integer(r.get("epochWidth")));
        String seq = pad(inputs.get("seq"), "seq", integer(r.get("sequenceWidth")));
        trace.add(entry("step", "pad", "epoch", epoch, "seq", seq));
        String iri = render((String) r.get("iriTemplate"), Map.of("namespace", namespace, "epoch", epoch, "seq", seq));
        return done(iri, trace, null);
    }

    private static String pad(Object v, String name, int width) {
        if (!(v instanceof Long || v instanceof Integer || v instanceof BigInteger)) {
            throw new MintException(MintError.MissingKeyComponent, name + " is missing or not an integer");
        }
        if (v instanceof BigInteger || ((Number) v).longValue() < 0) {
            throw new MintException(MintError.PositionOutOfRange, name + " " + v + " is outside 0.." + MAX_POSITION);
        }
        String digits = Long.toString(((Number) v).longValue());
        return "0".repeat(Math.max(0, width - digits.length())) + digits;
    }

    private Minted content(Map<String, ?> inputs, List<Map<String, Object>> trace) {
        if (!(inputs.get("canonicalizer") instanceof String canonicalizer) || Ucd.trimWhiteSpace(canonicalizer).isEmpty()) {
            throw new MintException(MintError.CanonicalizerNotDeclared,
                    "content-addressed minting needs a statement of which RDFC-1.0 implementation produced "
                            + "the bytes (identity-minting-specification.md §7, CA-1)");
        }
        if (!(inputs.get("canonicalNQuads") instanceof String nquads)) {
            throw new MintException(MintError.MissingKeyComponent, "canonicalNQuads is missing");
        }
        String value = digest(nquads.getBytes(StandardCharsets.UTF_8), map(r.get("digest")), trace, null, null, null);
        return done(render((String) r.get("iriTemplate"), Map.of("digest", value)), trace, canonicalizer);
    }
}
